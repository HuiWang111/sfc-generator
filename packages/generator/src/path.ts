import type { ElementNode, TemplateNode } from './types'
import { isElementNode } from './is'

export class Path {
  constructor(private _node: TemplateNode) {}

  get parent() {
    return this._node.parent
  }

  get node() {
    return this._node
  }

  private _find(
    node: TemplateNode,
    predicate: (node: TemplateNode) => boolean,
  ): TemplateNode | null {
    if (isElementNode(node) && node.children) {
      for (const child of node.children) {
        if (predicate(child)) {
          return child
        }
        else if (isElementNode(child)) {
          const result = this._find(child, predicate)
          if (result) {
            return result
          }
        }
      }
    }
    return null
  }

  replaceWith(node: TemplateNode) {
    const parent = this._node.parent as ElementNode | undefined
    if (parent && parent.children) {
      const index = parent.children.indexOf(this._node)
      if (index > -1) {
        parent.children[index] = node
      }
    }
    node.parent = parent
    this._node = node
  }

  remove() {
    const parent = this._node.parent as ElementNode | undefined
    if (parent && parent.children) {
      const index = parent.children.indexOf(this._node)
      if (index > -1) {
        parent.children.splice(index, 1)
      }
    }
  }

  skip() {
    if (isElementNode(this._node) && this._node.children) {
      this._node.children.forEach((child) => {
        child.__skip__ = true
      })
    }
  }

  getSibling(index: number) {
    const parent = this._node.parent as ElementNode | undefined
    if (parent && parent.children) {
      return parent.children[index]
    }
    return null
  }

  find(predicate: (node: TemplateNode) => boolean) {
    return this._find(this._node, predicate)
  }
}
