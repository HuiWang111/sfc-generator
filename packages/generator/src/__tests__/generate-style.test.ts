import { readFile } from 'node:fs/promises'
import { join } from 'node:path'
import { parseComponent } from '@vue/compiler-sfc'
import { generateStyle } from '../index'

describe('test generateStyle', () => {
  it('generateStyle should work', async () => {
    const sfc = await readFile(join(__dirname, 'components/form.vue'), 'utf-8')
    const sfcDescriptor = parseComponent(sfc)

    const code = generateStyle(sfcDescriptor.styles)
    expect(code).toMatchSnapshot()
  })
})
