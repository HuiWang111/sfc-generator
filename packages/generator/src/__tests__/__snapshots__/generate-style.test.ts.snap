// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test generateStyle generateStyle should work 1`] = `
"<style scoped>
.container {
  width: 100%;
  height: 100%;
}
</style>

<style>
.submit-container {
  width: 100%;
  height: 100%;
}
</style>"
`;
